rfc/custom-modules-v1.md

```diff
@@ -0,0 +1,456 @@
+	next.ServeHTTP(ctx.ResponseWriter(), ctx.Request())
+}
+
+func (m *MyModule) Module() core.ModuleInfo {
+	return core.ModuleInfo{
+		// This is the ID of your module, it must be unique
+		ID: myModuleID,
+		// The priority of your module, lower numbers are executed first
+		Priority: 1,
+		New: func() core.Module {
+			return &MyModule{}
+		},
+	}
+}
+
+// Interface guard
+var (
+	_ core.RouterMiddlewareHandler = (*MyModule)(nil)
+	_ core.EnginePreOriginHandler  = (*MyModule)(nil)
+	_ core.EnginePostOriginHandler = (*MyModule)(nil)
+	_ core.Provisioner             = (*MyModule)(nil)
+	_ core.Cleaner                 = (*MyModule)(nil)
+)
+```
+
+</details>
+
+## Proposal
+
+A developer can implement a custom module by creating a struct that implements one or more of the following interfaces:
+
+- `GatewayHooks`: Provides hooks for the gateway lifecycle, including request and response handling.
+- `SubgraphHooks`: Provides hooks for subgraph requests and responses.
+- `ApplicationHooks`: Provides hooks for the application lifecycle, including startup, shutdown, and error handling.
+- `TelemetryHooks`: Provides hooks for OpenTelemetry spans and metrics.
+- `OperationHooks`: Provides hooks for parsed, normalized, and planned GraphQL operations.
+- `Module`: Provides hooks for the module lifecycle, including provisioning and cleanup.
+
+```go
+type MyModule struct{}
+
+type Subgraph struct {
+	// The name of the subgraph
+	Name string
+	// The ID of the subgraph
+	ID string
+	// The URL of the subgraph
+	URL *url.URL
+	// The schema of the subgraph
+	Schema *graphql.Schema
+}
+
+type Graph struct {
+	// The name of the graph
+	Name string
+	// The ID of the graph
+	ID string
+	// Version of the graph
+	Version string
+	// Subgraphs of the graph
+	Subgraphs []*core.Subgraph
+	// GraphQL schema of the graph
+	Schema *graphql.Schema
+}
+
+type GatewayRequest struct {
+	// The original GraphQL request with all the information like query, variables, operation name, extensions etc.
+	Request *core.GraphQLRequest
+	// The parsed, normaliazed and planned operation with all the information like name, variables, type, document representation,
+	// client name version, uploaded files, plan, normalization, persisted operation etc.
+	Operation *core.Operation
+	// The active Telemetry instance
+	Telemetry *core.Telemetry
+	// The active graph
+	Graph *core.Graph
+	// The original HTTP request
+	Orignal *http.Request
+	// Logger for the gateway request
+	Logger *zap.Logger
+}
+
+type GatewayResponse struct {
+	// The original RouterRequest
+	Request *core.GatewayRequest
+	// The active Telemetry instance
+	Telemetry *core.Telemetry
+	// The final GraphQL response with all the information like data, errors, extensions etc.
+	// This is the response that will be sent to the client and can be manipulated or replaced
+	Response *core.GraphQLResponse
+	// The original HTTP response
+	Orignal *http.Response
+}
+
+type GatewayHooks interface {
```

Comment:
Gateway or Router?

Reply:
No strong objection, see my comment on the other thread.

Reply:
I think Gateway makes sense.